### PR TITLE
SES1251 - App crash on non alphanumeric RE-FIX

### PR DESCRIPTION
**Pull request by AL-Session**
### Contributor checklist

- [X] I have tested my contribution on these devices:
 * Virtual device W, Pixel 3a on Android 9 / API 28
 * Virtual device X, Pixel 3a on Android 14 / API 34
- [X] My contribution is fully baked and ready to be merged as is
- [X] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description
Searching messages and entering a non-alphanumeric character as the first character caused an crash & restart. This was due to the search being performed after each character entered, and any illegal character entered as the 1st character is replaced with a space - so in effect we are searching on the string `" "` which was causing the error.

Guards have been put in place to prevent search queries on strings of `" "` which eliminates the error.

Tested by entering a variety of non-alphanumeric characters as first search letter of message search both in regular user-to-user comms & in a group- all fine now.

### CODE REVIEW

File: app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt [hunk 1]
```diff
@@ -41,14 +48,19 @@ class GlobalSearchViewModel @Inject constructor(private val searchRepository: Se
         _queryText
                 .buffer(onBufferOverflow = BufferOverflow.DROP_OLDEST)
                 .mapLatest { query ->
+                    // Minimum search term is 2 characters
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt:51]: iOS lets us search 1 character.

It's not necessarily a good search, but you can give yourself a 1-char name, so you should be able to search that... and ideally a match of an entire word should be elevated.

File: app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt [hunk 2]
```diff
@@ -41,14 +48,19 @@ class GlobalSearchViewModel @Inject constructor(private val searchRepository: Se
...
                         SearchResult.EMPTY
                     } else {
                         // user input delay here in case we get a new query within a few hundred ms
                         // this coroutine will be cancelled and expensive query will not be run if typing quickly
                         // first query of 2 characters will be instant however
                         delay(300)
-                        val settableFuture = SettableFuture<SearchResult>()
+
+                        // If we already have a search running then stop it
+                        if (!settableFuture.isDone) { settableFuture.cancel(true) }
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt:61]: you don't need to check if the `Future` done, it's checked in `cancel`, the comment is not exactly true, as canceling this future would essentially stop listening for results, but won't stop any work being done.

Is it even necessary? `mapLatest` will prevent stale results clobbering newer ones.

File: app/src/main/java/org/thoughtcrime/securesms/search/SearchRepository.java [hunk 3]
```diff
@@ -123,23 +102,20 @@ public void query(@NonNull String query, @NonNull Callback<SearchResult> callbac
   }
 
   public void query(@NonNull String query, long threadId, @NonNull Callback<CursorList<MessageResult>> callback) {
-    if (TextUtils.isEmpty(query)) {
+    // If the sanitized search query is empty or less than 2 chars then abort the search
+    String cleanQuery = sanitizeQuery(query).trim();
+    if (cleanQuery.isEmpty() || cleanQuery.length() < 2) {
```
Comment (bemusementpark) [inline app/src/main/java/org/thoughtcrime/securesms/search/SearchRepository.java:107]: a bit tautological, `length < 2` would cover both cases, but maybe the `isEmpty` is preferred?

File: app/src/main/java/org/thoughtcrime/securesms/home/HomeActivity.kt [hunk 4]
```diff
@@ -82,8 +89,10 @@ import org.thoughtcrime.securesms.util.disableClipping
 import org.thoughtcrime.securesms.util.push
 import org.thoughtcrime.securesms.util.show
 import org.thoughtcrime.securesms.util.themeState
+
 import java.io.IOException
 import java.util.Locale
+
```
[the remark this answers is not in the document]
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/home/HomeActivity.kt:95]: Nah, I just tend to prefer lines between import groups, but I can revert :) / Done

File: app/src/main/java/org/thoughtcrime/securesms/search/SearchRepository.java (hunk 3 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/search/SearchRepository.java:107]: Adjusted so only using sanitize -> trim -> isEmpty

File: app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt (hunk 2 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt:61]: Cool, I didn't know `mapLatest` did that - have reverted `settableFuture` back to being a val defined at point of use

File: app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt (hunk 1 above)
Comment (AL-Session) [reply app/src/main/java/org/thoughtcrime/securesms/home/search/GlobalSearchViewModel.kt:51]: Yeah, fair enough - have allowed 1 char searches and it does prioritise contacts & groups over individual messages.

**Review by bemusementpark — APPROVED** [review]
nice!

**Pull request merged**